### PR TITLE
Simplify TPP

### PR DESCRIPTION
**Pull request by subhasisb**
#### Describe Bug or Feature

TPP networking layer had a lot of difficult to maintain and complicated code to support seamless backup transmission circuits. This functionality of a backup circuit is necessary for failover and resilience, but the way it is implemented in TPP currently is an overkill. TPP attempts to continue transmitting and receiving data for a stream seamlessly over multiple circuits. After the addition of features in PBS mom where jobs are no longer killed if network is temporarily down, such sophistication and complication is not required. When a connection breaks, TPP can close streams over those connections (unlike before) and the the users (mom, server etc) can re-open the required streams over the backup circuits (in lieu of never seeing the stream going down before)

#### Describe Your Change

The TPP framework had a lot of code to handle data for the same stream being transmitted over multiple circuits, and hence needed to sequence packets, keep packets at source for retransmission, and send and receive acks for packets. This was duplicate function on top of the similar functionality provided by TCP (which TPP uses underneath anyway) and was required only because the design goal was to never close the stream if an alternate circuit was available. However, since there is no practical impact/harm of closing the stream and allowing the application layer (mom, server) to re-open those same streams as and when required, we can remove the code related to packet acks retries etc.

Since the layer will now directly only use the TCP functionality, rpp_retry and rpp_highwater has no effect on the TPP streams. These values, if set, will be silently ignored (qmgr will not error if admin attempts to set these values - however a log warning that these switches are deprecated will be printed in the server logs)

Another significant change done is to the way the pbs_comm router works. Eariler we had only a mutex to go over the shared avltree tables that mapped destination IPs to comm routers. This synchronization mechanism is now changed to use pthread rwlocks where multiple readers can simultaneously traverse the avltrees, without needing to wait for each other. Minor changes were done to the avltree implementation to support the same as well.

While we are simplifying things, the logging related mutexes and fork handlers were simplified as well. The atfork() handlers do not need to do all that it was doing - primarily because, we know that only the APP thread (the pbs daemons thread) call fork, ever. The TPP thread would never called fork. When fork() is called, only the calling thread is duplicated in the child process, but the entire heap is copied over. This means locks, if held by the TPP thread would be copied over to the child process in "locked" state and can never be unlocked (because the TPP thread is dead, and nobody else can unlock a lock that is acquired by a different thread). The atfork handlers exist were there to allow the APP thread to acquire the mutex, so that the child can actually safely unlock it in the APP thread. However, no TPP functionality is called after a fork in the child, so only the logging needs to be handled. Thus only the pbs_log.c will have atfork handlers.

The tpp_log() function is also simplified to no longer need a TLS based buffer, now using only a stack based buffer. This also means that there is no need to register a logger function with the tpp layer.  

#### Attach Test and Valgrind Logs/Output

Description: Tests from given sources on platforms SUSE15, CENTOS8
Platforms: SUSE15,CENTOS8
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|5306|1927|1|1|0|0|1925|

Description: Tests from given sources on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8, CENTOS7, RHEL7, SLES12, SLES15, WIN2K16
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|5310|3842|5|0|0|0|3837|

The failures are known test/product issues for which bugs have been raised earlier.

### Hellgrind thread race detection tests
PBS_CONF_FILE=$conf_file valgrind --tool=helgrind --log-file=/tmp/mom_hellgrind.txt $i ${PBS_EXEC}/sbin/pbs_mom -N > /tmp/mom$i.out 2>&1 &
[mom_hellgrind.txt](https://github.com/openpbs/openpbs/files/5693693/mom_hellgrind.txt)

/usr/bin/valgrind.bin --tool=helgrind --log-file=./comm_hellgrind.txt pbs_comm -t 20
[comm_hellgrind.txt](https://github.com/openpbs/openpbs/files/5693724/comm_hellgrind.txt)

 valgrind --tool=helgrind --log-file=./server_hellgrind.txt pbs_server.bin
[server_hellgrind.txt](https://github.com/openpbs/openpbs/files/5693729/server_hellgrind.txt)

### Valgring memory error detection run results
[val_comm_74873.txt](https://github.com/openpbs/openpbs/files/5719336/val_comm_74873.txt)
[val_mom_79435.txt](https://github.com/openpbs/openpbs/files/5719337/val_mom_79435.txt)
[val_svr_79453.txt](https://github.com/openpbs/openpbs/files/5719338/val_svr_79453.txt)

### Failover tests output
Steps:
1. Configure a failover setup with a primary pbs server host/container, a secondary pbs server host/container and a separate mom container. The primary and the secondary hosts/containers share the PBS_HOME over a shared FS
2. Submit a long running job that starts running on the mom
3. Kill only the primary server process. Observe, that job is not killed, mom reconnects to secondary pbs server, but via the primary comm (primary comm keeps logging)
4. Reverse the test: Bring back primary server and kill only the primary comm. Observe that job is not killed, mom drops connection for a moment but reconnects to server via the secondary comm and continues normally
5. Kill the secondary comm as well. Notice that mom loses connections to server, but does not kill job.
6. Bring up primary/secondary comm (any one): Notice that mom reconnects back and works normally
[server.log](https://github.com/openpbs/openpbs/files/5693942/server.log)
[comm.log](https://github.com/openpbs/openpbs/files/5693943/comm.log)
[mom.log](https://github.com/openpbs/openpbs/files/5693946/mom.log)

### TPP regression automation tests output

[Execution_logs_TestTPP_2moms_2comms.txt](https://github.com/openpbs/openpbs/files/5719134/Execution_logs_TestTPP_2moms_2comms.txt)
[iso_pool2.txt](https://github.com/openpbs/openpbs/files/5719135/iso_pool2.txt)
[TestTPP_2moms.txt](https://github.com/openpbs/openpbs/files/5719137/TestTPP_2moms.txt)
[TestTPP_3moms_3comms.txt](https://github.com/openpbs/openpbs/files/5719138/TestTPP_3moms_3comms.txt)
[TestTPP_3rdfromlast.txt](https://github.com/openpbs/openpbs/files/5719139/TestTPP_3rdfromlast.txt)
[TestTPP_5comms.txt](https://github.com/openpbs/openpbs/files/5719140/TestTPP_5comms.txt)
[TestTPP_client_only.txt](https://github.com/openpbs/openpbs/files/5719407/TestTPP_client_only.txt)
[TestTPP_mom_nonserver_2.txt](https://github.com/openpbs/openpbs/files/5719408/TestTPP_mom_nonserver_2.txt)
[TestTPP_nonserver1.txt](https://github.com/openpbs/openpbs/files/5719409/TestTPP_nonserver1.txt)
[comm_with_vnode_insertion.txt](https://github.com/openpbs/openpbs/files/5719411/comm_with_vnode_insertion.txt)

### CODE REVIEW

**Review by hirenvadalia — CHANGES_REQUESTED** [review]

File: src/lib/Libtpp/tpp_util.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_util.c:200]: We can't - check out how log_errf() etc do the vsprintf inside and then call log_event() or log_record(). Point is, passing the varargs to log_eventf() is not straightforward, I think

File: src/lib/Libtpp/tpp_util.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_util.c:2393]: yes, of course - good catch - not sure what i was thinking

File: src/lib/Libtpp/tpp_client.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_client.c:3081]: yes - never noticed that carefully - thanks

File: src/lib/Libtpp/tpp_client.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_client.c:951]: yes, thanks

File: src/lib/Liblog/pbs_log.c
Comment (subhasisb) [reply src/lib/Liblog/pbs_log.c]: reverted back to original

File: src/lib/Libtpp/tpp_client.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_client.c:278]: yeah i will rename to tpp_terminated_in_child

File: src/lib/Libtpp/tpp_client.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_client.c:2373]: yeah it is a vestige from older impl - now we do not have part packets, so removed this - thanks for noting

File: src/lib/Libtpp/tpp_platform.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_platform.c:145]: :) darn - thanks for noting. i had removed the implementation of tpp_get_logbuf() anyway - so windows builds must have failed.

File: src/lib/Libtpp/tpp_router.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_router.c:1123]: I think we are passing &dhdr and dhdr is itself a pointer. This means the decrypt_data is supposed to "allocate" the buffer and set the pointer to the NEW buffer and return the length in len_out.

Indeed from your own documentation: Cleanup: Returned data_out should be freed using free() when no more required, as it will be allocated heap memory. So, we are missing the free - need to chat

File: src/lib/Libutil/avltree.c
Comment (subhasisb) [reply src/lib/Libutil/avltree.c:129]: sure we can.

File: src/lib/Liblog/pbs_log.c
Comment (hirenvadalia) [inline src/lib/Liblog/pbs_log.c:1052]: I think you will need to add below code here for some platform else it will give compile error:
```c
#else
     return;
#endif
```

Comment (subhasisb) [reply src/lib/Liblog/pbs_log.c:1052]: its a void function, so no need to add return

File: src/resmom/mom_main.c
Comment (agrawalravi90) [inline src/resmom/mom_main.c:8200]: comment seems unnecessary

File: src/lib/Libutil/avltree.c
Comment (agrawalravi90) [inline src/lib/Libutil/avltree.c:137]: Oh, I was not aware of this. Thanks for adding the comment

File: src/lib/Libutil/avltree.c
Comment (agrawalravi90) [inline src/lib/Libutil/avltree.c:167]: static?

File: src/lib/Libutil/avltree.c
Comment (agrawalravi90) [inline src/lib/Libutil/avltree.c:168]: Will pthread_getthreadid_np() not work?

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]

File: src/include/tpp.h
Comment (subhasisb) [reply src/include/tpp.h:112]: I need to raise a design doc to say deprecated, but for now these qmgr's will be accepted but will be silently ignored

File: src/lib/Liblog/pbs_log.c
Comment (subhasisb) [reply src/lib/Liblog/pbs_log.c:446]: We need to get the timestamp in log_record() itself to know if we need to switch logs, and then again in log_record_inner() to put in the log record. I was thinking we don't need to call it twice, especially since log_record_inner() is a static method and not an interface.

File: src/lib/Liblog/pbs_log.c
Comment (subhasisb) [reply src/lib/Liblog/pbs_log.c:999]: We could use PTHREAD_MUTEX_RECURSIVE of course, however, PTHREAD_MUTEX_RECURSIVE locks do not play well across forks over atfork handlers, so we are stuck to using non-recursive locks

File: src/lib/Libtpp/tpp_client.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_client.c]: We can potentially, but tpp_log is really a wrapper over log_event, it adds a few more information like thread id etc, that log_event does not understand - and adding it in each line we call is going to be cumbersome for the caller.

File: src/lib/Libtpp/tpp_client.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_client.c:402]: no - actually it will create a new chunk and add the end of the packet referenced by pkt

File: src/lib/Libtpp/tpp_client.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_client.c]: For now the thread related code is nicely tucked inside TPP and it uses the normal log_event() eventually

File: src/lib/Libtpp/tpp_client.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_client.c]: The packet structure takes care of it

File: src/lib/Libutil/avltree.c
Comment (subhasisb) [reply src/lib/Libutil/avltree.c:173]: I recall pthread_getthreadid_np() returns a posix id for the thread. What i need is an integer to index into an array, hence using an internal number

Comment (agrawalravi90) [reply src/lib/Libutil/avltree.c:173]: Looks like pthread_getthreadid_np() does return an integer, but it's non-posix, which is what the 'np' stands for. So nevermind.

Comment (subhasisb) [reply src/lib/Libutil/avltree.c:173]: yes np is for that, but also the integer is its own number (i meant that when i said posix, i should have said internal number) which i cannot use to index into my array

**Review by hirenvadalia — APPROVED** [review]

**Comment by subhasisb** [comment]
Do not merge it yet. I am running some more tests and waiting for windows builds to succeed before we can merge it.

File: src/lib/Libtpp/tpp_router.c
Comment (agrawalravi90) [inline src/lib/Libtpp/tpp_router.c:321]: where did we acquire this lock? it might be better to unlock in the same function that locks it, if possible.

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]
Minor comment

File: src/lib/Libtpp/tpp_router.c
Comment (subhasisb) [reply src/lib/Libtpp/tpp_router.c:321]: The lock is acquired in the calling function, please see the header

File: src/include/avltree.h
Comment (subhasisb) [inline src/include/avltree.h:82]: blocking

**Review by subhasisb — COMMENTED** [review]
blocking for now, till i complete tests

**Comment by subhasisb** [comment]
All tests and builds done. This is now ready to merge

**Review by agrawalravi90 — APPROVED** [review]

**Pull request merged**
